Touchscreen SR changes

BenZeng04 opened the pull request:
It is quite very clear that there are an incredible amount of things wrong with the pp computation with touchscreen, so I have developed SR changes for touchscreen that go into a very huge variety of skills.

Here is a summary of what will happen:
**Aim Changes:**
Angle is taken into account in a different way than it is currently. Back and forths get a very slight nerf while anything wider than back and forths will get a slight buff as angle increases. This buff will not increase linearly, as a 90 degree angle jump is arguably the same difficulty as a 180 degree angle one - you will hit one note with one hand, the other two with another.

Fast stacked notes, such as triples often require moving your other hand to press the X key, or TapX. This results in not being able to use this hand for a short amount of time, making aiming significantly harder for the next few notes. Notes after fast enough streams that are not stacked will get an aim boost.

**Speed Changes:**
Streams are currently extremely difficult with touchscreen, spaced streams being nearly impossible without simply singletapping all the notes in the stream. Not much to say about this change, but streams will be getting a big overall buff, as long as they are spaced enough, fast enough to be unsingletappable, and wide angle to prevent cheesing with multiple hands.

**PP Changes:**
See [here](https://github.com/ppy/osu-performance/issues/81)

**Misc. Changes:**
Strain is calculated by adding current strain to note **before** the previous note, not the previous note itself. This puts a small nerf to large spacing diff spikes.

It is worth it to mention that touchscreen pp will never go above non-touchscreen due to a cap. This is to prevent manually detecting touchscreen for additional pp on stream maps.

BenZeng04 committed:
Create TouchAim.cs

BenZeng04 committed:
Update TouchAim.cs

BenZeng04 committed:
added descriptions

BenZeng04 committed:
added descriptions

BenZeng04 committed:
Create TouchSpeed.cs

BenZeng04 committed:
Update TouchSpeed.cs

BenZeng04 committed:
Update TouchAim.cs

BenZeng04 committed:
Update TouchSpeed.cs

BenZeng04 committed:
Update TouchSpeed.cs

BenZeng04 committed:
Adjust values

BenZeng04 committed:
Update TouchSpeed.cs

BenZeng04 committed:
Adjust touchscreen nerf original values

BenZeng04 committed:
Update TouchSpeed.cs

BenZeng04 committed:
Update OsuPerformanceCalculator.cs

TDRework committed:
replaced everything

TDRework committed:
Revert accuracy change

TDRework committed:
Balanced old values with new values

BenZeng04 committed:
Merge branch 'master' into master

TDRework committed:
Resolve Issues

TDRework committed:
Update Skill.cs

TDRework committed:
Merge branch 'master' of https://github.com/yeahbennou/osu

TDRework committed:
Update Skill.cs

TDRework committed:
trimmed white space

TDRework committed:
Trim white space

because i have to

TDRework committed:
trim white space!!!

TDRework committed:
Trim white space again

>:(

TDRework committed:
Update TouchAim.cs

TDRework committed:
Update ManiaDifficultyCalculatorTest.cs

yeahbennou commented:
[For some apparent reason, a mania difficulty calculation test failed so I adjusted the test values](https://github.com/ppy/osu/pull/4326/commits/40fb341457cd7e98d7e17526229ecd045f6210fa)

TDRework committed:
Small adjustments to TD calculations

Commit:
Revert small adjustments

Commit:
Nerfed overall buff to return less extreme values

Commit:
Fixed Stacked streams being worth less with touchscreen than without

Commit:
Adjusted values

BenZeng04 committed:
Merge branch 'master' into master

yeahbennou commented:
Example Values


[EbonSol](https://hastebin.com/igubemunuv.sql)
[itsamemarioo](https://hastebin.com/loqenikori.sql)
[freedomdiver](https://hastebin.com/loqejaluqi.css)
[yeahbennou](https://hastebin.com/ugibuguwal.sql)

Commit:
Update ManiaDifficultyCalculatorTest.cs

Commit:
Merge branch 'master' of https://github.com/yeahbennou/osu

smoogipoo commented on osu.Game/Screens/Play/PlayerSettings/PlaybackSettings.cs:
```diff
@@ -72,7 +72,7 @@ protected override void LoadComplete()
             rateSlider.Bindable.ValueChanged += multiplier => AdjustableClock.Rate = clockRate * multiplier.NewValue;
 
-            rateSlider.Bindable.BindValueChanged(multiplier => multiplierText.Text = $"{multiplier.NewValue:0.0}x", true);
+            rateSlider.Bindable.BindValueChanged(multiplier => multiplierText.Text = $"{multiplier:0.0}x", true);
```
Revert changes to this file.

yeahbennou replied:
Done.

smoogipoo commented on osu.Game/Rulesets/Difficulty/Skills/Skill.cs:
```diff
@@ -95,7 +114,6 @@ public double DifficultyValue()
                 difficulty += strain * weight;
                 weight *= DecayWeight;
             }
-
```
Revert

yeahbennou replied:
Done.

smoogipoo commented on osu.Game/Rulesets/Difficulty/Skills/Skill.cs:
```diff
@@ -58,6 +61,21 @@ public void Process(DifficultyHitObject current)
             Previous.Push(current);
         }
 
+        public void TouchProcess(DifficultyHitObject current)
```
This code affects all difficulty calculators and should not be in here.

yeahbennou replied:
Issue has been resolved as a result of reverting strain calculation change completely.

smoogipoo commented on osu.Game/Rulesets/Difficulty/Skills/Skill.cs:
```diff
@@ -58,6 +61,21 @@ public void Process(DifficultyHitObject current)
 
+        public void TouchProcess(DifficultyHitObject current)
+        {
+            var dummy = prevDummyStrain;
```
Words like `dummy` are generally intended to be used for discarded values. This doesn't seem like a discarded value.

Please rename it to something more descriptive.

yeahbennou replied:
Issue has been resolved as a result of reverting strain calculation change completely.

smoogipoo commented on osu.Game/Rulesets/Difficulty/Skills/Skill.cs:
```diff
@@ -58,6 +61,21 @@ public void Process(DifficultyHitObject current)
+            prevDummyStrain = previous * strainDecay(current.DeltaTime);
+            prevDummyStrain += StrainValueOf(current) * SkillMultiplier;
+            currentStrain = (prevDummyStrain + dummyStrain) / 2;
+            previous = dummy;
```
I'd like comments to be added to this method explaining what this process does. I.e. why it's using the last-last object strain, and why the averaging is done above.

yeahbennou replied:
The old values used to only use the last-last object strain instead of an average due to how touchscreen generally used two cursors. However, after further discussion with touchscreen players, we have found that it would make the most sense to average it to dilute the overall results. After further discussion, we have found that it would make the most sense to revert this completely. This would in turn, resolve many of the issues that derive from this.

smoogipoo replied:
If it were something done entirely inside `TouchSpeed`/`TouchAim`, I could reconsider, but this is fundamentally changing the mathematical basis of how strains are calculated from one strain to the next rather than changing the strain of a particular hitobject.

For example if you wanted to enforce something like "2 cursors" you could divide the screen into two halves and give a hitobject a strain of 0 if it occured in the opposite half as the last hitobject (or last-last object, or last-last-last object). That would be an implementation of this that I would be fine with, which wouldn't require changing the underlying `apply decay -> apply strain value -> get maximum` process.

Commit:
Add a cap so TD will never exceed non-TD

smoogipoo commented on osu.Game/Rulesets/Difficulty/Skills/Skill.cs:
```diff
@@ -41,7 +40,11 @@ public abstract class Skill
+        private double prevDummyStrain = 1; // Balancing touchscreen strain so that effects are smaller.
+        private double previous = 0;
         private double currentSectionPeak = 1; // We also keep track of the peak strain level in the current section.
+        private double dummySectionPeak = 1; // Balancing touchscreen strain so that effects are smaller.
```
These comments aren't really helpful, see my other comment.

yeahbennou replied:
Issue has been resolved as a result of reverting strain calculation change completely.

BenZeng04 committed:
Merge branch 'master' into master

smoogipoo commented on osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs:
```diff
@@ -87,11 +86,19 @@ private DifficultyAttributes calculate(IBeatmap beatmap, Mod[] mods, double cloc
                 foreach (Skill s in skills)
-                    s.Process(h);
+                {
+                    if (counted > 1)
```
Instead of this, create a virtual method in `Skill` called `ComputeStrain()`. Have `Skill.Process()` call this method, and have `TouchSkill` derive this method with your implementation.

Then you'd have your `TouchSpeed` and `TouchAim` skills derive `TouchSkill`.

With that being said, I'm like 97% sure this sort of overriding is definitely _not_ what we want, and this looks like throwing darts at the wall. It doesn't make logical sense to me, and the absolute necessity of it is not described anywhere.

@Tom94 please see if you agree with my assertion.

yeahbennou replied:
Issue has been resolved as a result of reverting strain calculation completely.

smoogipoo commented on osu.Game.Rulesets.Mania.Tests/ManiaDifficultyCalculatorTest.cs:
```diff
@@ -13,7 +13,7 @@ public class ManiaDifficultyCalculatorTest : DifficultyCalculatorTest
         protected override string ResourceAssembly => "osu.Game.Rulesets.Mania";
 
-        [TestCase(2.3683365342338796d, "diffcalc-test")]
+        [TestCase(2.9183460420403056d, "diffcalc-test")]
```
Revert, if this changes in a non-mania-related change then your code is broken.

yeahbennou replied:
Done

Commit:
Resolving and reverting changes as a request

Commit:
Modifying values to compensate for reverting of strain calculation

yeahbennou commented:
Misc. Changes have been reverted

Commit:
Added descriptions

Commit:
Trimmed white space

BenZeng04 committed:
Merge branch 'master' into master

Commit:
Modified Speed pp calculation

rawSpeed is no longer nerfed. To compensate, the speedvalue given for large spacing jumps has been reduced. This effectively nerfs fast jumps and gives more emphasis to actual streams in speed for touchscreen.

Commit:
Made minor adjustments to values

More specifically, a smaller overall nerf to rawaim and a bigger nerf to back and forths, as well as a bigger nerf to fast jumps.

yeahbennou commented:
Now that quite a bit has changed, here is an updated version of what will happen:

**Aim Changes:**
It has been quite agreed upon that the original nerf - Aim Stars ^ 0.8 was quite harsh and led to nerfs of over 500pp in a single play. I have reduced this nerf to 0.86 instead of 0.8. Along with this, angle is taken into account in a very different way than it is currently. The current system buffs anything over 60 degrees which increases as angle increases, which does not make sense from the perspective of touchscreen. Instead, back and forths will get a nerf, while anything wider than around 20 degrees will get a buff as angle increases, until around 90 degrees, where the buff will stop increasing as much. 

Fast stacked notes, such as triples often require moving your other hand to press the X key, or TapX. This results in not being able to use this hand for a short amount of time, making aiming significantly harder for the next few notes. Notes after fast enough streams that are not stacked will get an aim boost.

**Speed Changes:**
Streams are currently extremely difficult with touchscreen, spaced streams being nearly impossible with TapX, forcing you to singletap all the notes in the stream. Not much to say about this change, but streams will be getting a big overall buff, as long as they are spaced enough, fast enough to be unsingletappable, and wide angle to prevent cheesing with multiple hands. 

Along with the stream buff, fast jumps are getting a nerf for speed pp, which means that high spacing will no longer award as much of a speed pp bonus as it does now. This is due to how high bpm DT maps such as ILY, Make a Move, and Unknown Title currently award significantly more pp than they should, relative to other touchscreen plays, simply due to the presence of fast jumps inflating the speed pp value. This will not nerf spaced streams at all, as the current stream buff will give an overwhelmingly bigger buff than this change would nerf to spaced streams.

**PP Changes:**
See [here](https://github.com/ppy/osu-performance/issues/81)

Commit:
Adjusted values

for balancing

BenZeng04 committed:
Merge branch 'master' into master

Commit:
Angle bonus for fast jumps now scales speed value rather than distance

Commit:
Merge branch 'master' of https://github.com/yeahbennou/osu

BenZeng04 committed:
Merge branch 'master' into master

Commit:
initialize angle and jump as a double instead of var

Commit:
slight modification

BenZeng04 committed:
Merge branch 'master' into master

Commit:
Modified angle bonuses for Aim and Speed

BenZeng04 committed:
Merge branch 'master' into master

Commit:
Slight modifications to values

Commit:
Slightly modified values

Commit:
cleaned up unneccesary indents

BenZeng04 committed:
Merge branch 'master' into master

BenZeng04 committed:
Merge branch 'master' into master

BenZeng04 committed:
Reverted arbitrary changes to Skill file

BenZeng04 committed:
Reverted arbitrary changes

Commit:
Revert flashlight bonus to originally proposed bonuses

BenZeng04 committed:
Merge branch 'master' into master

BenZeng04 committed:
Merge branch 'master' into master

peppy commented:
Please stop merging master.

Commit:
Modified values to return sane results for extremely sharp angles

yeahbennou commented:
Closing for now to work on a much more in depth version of the aim rebalance which focuses on something more than just angle.
